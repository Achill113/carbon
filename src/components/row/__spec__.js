import React from 'react';
import TestUtils from 'react/lib/ReactTestUtils';
import Row from './index';

describe('Row', () => {
  it('failing test demo', () => {
    expect(1).toEqual(2);
  });
});
