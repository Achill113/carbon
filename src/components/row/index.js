import React from 'react';

class Row extends React.Component {

  buildColumns = () => {
    var columns = [];

    if (this.props.children.length) {
      this.props.children.forEach((child, index) => {
        columns.push(this.buildColumn(child, index));
      });
    } else {
      columns.push(this.buildColumn(this.props.children, 0));
    }

    return columns;
  }

  buildColumn = (child, key) => {
    var columnClass = "ui-row__column";

    if (child.props.columnClasses) {
      columnClass += " " + child.props.columnClasses;
    }

    if (child.props.columnOffset) {
      columnClass += " ui-row__column--offset-" + child.props.columnOffset;
    }

    if (child.props.columnSpan) {
      columnClass += " ui-row__column--span-" + child.props.columnSpan;
    }

    return (
      <div key={ key } className={ columnClass }>
        { child }
      </div>
    );
  }

  /**
   * Renders the component.
   *
   * @method render
   */
  render() {
    var mainClasses = "ui-row";

    if (this.props.columns) {
      mainClasses += " ui-row--columns-" + this.props.columns;
    } else {
      mainClasses += " ui-row--columns-" + this.props.children.length;
    }

    return (
      <div className={ mainClasses }>
        { this.buildColumns() }
      </div>
    );
  }

};

export default Row;
