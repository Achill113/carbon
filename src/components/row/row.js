import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { compact } from 'lodash';
import Immutable from 'immutable';
import Column from './column';
import Logger from './../../utils/logger';

/**
 * A row widget.
 *
 * This is a standalone row widget used for layout; for table rows use the table-row widget.
 *
 * == How to use a Row in a component:
 *
 * In your file
 *
 *   import Row from 'carbon/lib/components/row';
 *
 * To render the Row:
 *
 *   <Row />
 *
 * @class Row
 * @constructor
 */
class Row extends React.Component {

  static propTypes = {

    /**
     * The elements to be rendered in the row
     *
     * @property children
     * @type {Object | Array}
     */
    children: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.object
    ]),

    /**
     * Custom className
     *
     * @property className
     * @type {String}
     */
    className: PropTypes.string,

    /**
     * Pass a custom value for the gutter
     * (extra-small, small, medium, large or extra-large)
     *
     * @property gutter
     * @type {String}
     */
    gutter: PropTypes.string,

    /**
     * Show a divide between columns
     *
     * @property columnDivide
     * @type {String}
     */
    columnDivide: PropTypes.bool,

    /**
     * Manually define number of columns
     *
     * @property columns
     * @type {String}
     */
    columns: PropTypes.string
  }

  static defaultProps = {
    gutter: 'medium'
  };

  /**
   * Builds row columns from the children object fields
   *
   * @method buildColumns
   * @return {Array} array of built columns
   */
  buildColumns = () => {
    return React.Children.map(this.props.children, (child) => {
      return React.cloneElement(
        child, {
          columnClasses: this.props.columnClasses,
          columnDivide: this.props.columnDivide
        },
        child.props.children
      );
    });
  }

  /**
   * Main Class getter
   *
   * @method mainClasses
   * @return {String} Main className
   */
  get mainClasses() {
    let columns = 1;

    if (this.props.columns) {
      columns = this.props.columns;
    } else if (this.props.children && this.props.children.constructor === Array) {
      columns = compact(this.props.children).length;
    } else if (Immutable.Iterable.isIterable(this.props.children)) {
      columns = this.props.children.size;
    }

    return classNames(
      'carbon-row',
      `carbon-row--gutter-${this.props.gutter}`,
      this.props.className,
      `carbon-row--columns-${columns}`
    );
  }

  /**
   * Renders the component.
   *
   * @method render
   * @return {Object} JSX
   */
  render() {
    return (
      <div className={ this.mainClasses }>
        { this.buildColumns() }
      </div>
    );
  }
}

export default Row;
export {
  Row,
  Column
};
