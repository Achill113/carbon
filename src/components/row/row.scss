@import 'general';

$max-columns: 12 !default;
$sizes: ("extra-small", $grid-margin-extra-small),
        ("small", $grid-margin-small),
        ("medium-small", $grid-margin-medium-small),
        ("medium", $grid-margin-medium),
        ("medium-large", $grid-margin-medium-large),
        ("large", $grid-margin-large),
        ("extra-large", $grid-margin-extra-large);

.carbon-row {
  clear: both;
  position: relative;

  @each $name, $size in $sizes {
    &.carbon-row--gutter-#{$name} {
      margin-bottom: -($size);
      margin-left: -($size);

      > .carbon-row__column {
        margin-bottom: $size;
        padding-left: $size;
      }
    }
  }

  .carbon-row__column {
    box-sizing: border-box;
    float: left;
    min-height: 1px;
    width: 100%;

    &.carbon-row__column--align-center,
    &.carbon-row__column--align-middle {
      text-align: center;
    }

    &.carbon-row__column--align-right {
      text-align: right;
    }
  }

  @for $i from 2 through $max-columns {
    $width: 100% / $i;

    &.carbon-row--columns-#{$i} {
      > .carbon-row__column {
        width: $width;

        &:nth-child(#{$i}n+1) {
          clear: both;
        }
      }

      // span columns:
      @for $y from 2 through $i {
        @if $y != $i {
          > .carbon-row__column--span-#{$y} {
            width: (100% / $i) * $y;
          }
        }
      }

      // offset columns:
      @for $y from 1 through $i {
        @if $y != $i {
          > .carbon-row__column--offset-#{$y} {
            margin-left: $width * $y;
          }
        }
      }
    }
  }

  // Clearfix:
  &:before,
  &:after {
    content: " ";
    display: table;
  }
  &:after {
    clear: both;
  }
}
