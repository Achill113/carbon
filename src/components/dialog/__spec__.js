import React from 'react';
import TestUtils from 'react-dom/test-utils';
import { mount, shallow } from 'enzyme';
import Bowser from 'bowser';
import Dialog from './dialog';
import Button from './../button';
import { Row, Column } from './../row';
import { elementsTagTest, rootTagTest } from '../../utils/helpers/tags/tags-specs';

describe('Dialog', () => {
  let instance, onCancel;

  beforeEach(() => {
    onCancel = jasmine.createSpy('cancel');
  });

  describe('Lifecycle functions', () => {
    describe('componentDidMount', () => {
      describe('when dialog is open', () => {
        it('centers the dialog', () => {
          instance = TestUtils.renderIntoDocument(
            <Dialog open onCancel={ onCancel } />
          );
          spyOn(instance, 'centerDialog');
          instance.componentDidMount();
          expect(instance.centerDialog).toHaveBeenCalled();
        });
      });

      describe('when dialog is closed', () => {
        it('does not center the dialog', () => {
          instance = TestUtils.renderIntoDocument(
            <Dialog open={ false } onCancel={ onCancel } />
          );
          spyOn(instance, 'centerDialog');
          instance.componentDidMount();
          expect(instance.centerDialog).not.toHaveBeenCalled();
        });
      });
    });

    describe('componentDidUpdate', () => {
      describe('when the dialog is open', () => {
        beforeEach(() => {
          instance = TestUtils.renderIntoDocument(
            <Dialog open onCancel={ onCancel } />
          );
        });

        it('centers the dialog', () => {
          spyOn(instance, 'centerDialog');
          instance.componentDidUpdate();
          expect(instance.centerDialog).toHaveBeenCalled();
        });

        it('sets up event listeners to resize and close the dialog', () => {
          const spy = spyOn(window, 'addEventListener');
          instance.componentDidUpdate();
          expect(spy.calls.count()).toEqual(2);
          expect(window.addEventListener).toHaveBeenCalledWith('resize', instance.centerDialog);
          expect(window.addEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
        });

        describe('when the dialog is already listening', () => {
          it('does not set up event listeners', () => {
            const spy = spyOn(window, 'addEventListener');
            instance.listening = true;
            instance.componentDidUpdate();
            expect(spy.calls.count()).toEqual(0);
            expect(window.addEventListener).not.toHaveBeenCalled();
            expect(window.addEventListener).not.toHaveBeenCalled();
          });
        });
      });

      describe('when the dialog is closed', () => {
        beforeEach(() => {
          instance = TestUtils.renderIntoDocument(
            <Dialog open={ false } onCancel={ onCancel } />
          );
        });

        it('removes event listeners for resize and closing', () => {
          const spy = spyOn(window, 'removeEventListener');
          instance.componentDidUpdate();
          expect(spy.calls.count()).toEqual(2);
          expect(window.removeEventListener).toHaveBeenCalledWith('resize', instance.centerDialog);
          expect(window.removeEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
        });
      });
    });
  });

  describe('centerDialog', () => {
    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Dialog open onCancel={ onCancel } />
      );
    });

    describe('when dialog is lower than 20px', () => {
      it('sets top position to the correct value', () => {
        instance.centerDialog();
        expect(instance._dialog.style.top).toEqual('150px');
      });
    });

    describe('when dialog is higher than 20px', () => {
      it('sets top position to 20px', () => {
        instance._dialog = {
          style: {},
          offsetHeight: 261
        };
        instance.centerDialog();
        expect(instance._dialog.style.top).toEqual('20px');
      });
    });

    describe('when dialog is less than 20px from the side', () => {
      it('sets top position to 20px', () => {
        instance._dialog = {
          style: {},
          offsetWidth: 361
        };
        instance.centerDialog();
        expect(instance._dialog.style.left).toEqual('20px');
      });
    });

    describe('when ios', () => {
      it('does not remove page y offset', () => {
        Bowser.ios = true;
        instance.centerDialog();
        expect(instance._dialog.style.top).toEqual('150px');
      });
    });
  });

  describe('dialogTitle', () => {
    describe('when a props title is passed', () => {
      it('sets a dialog header', () => {
        const wrapper = mount(
          <Dialog
            onCancel={ onCancel }
            open
            title='Dialog title'
            subtitle='Dialog subtitle'
          />
        ),
            title = wrapper.find('[data-element="title"]'),
            subtitle = wrapper.find('[data-element="subtitle"]');

        expect(title.props().children).toEqual('Dialog title');
        expect(subtitle.props().children).toEqual('Dialog subtitle');
      });
    });

    describe('when a props object title is passed', () => {
      it('sets a dialog header', () => {
        const wrapper = mount(
          <Dialog
            onCancel={ onCancel }
            open
            title={
              <Row>
                <Column>Row1</Column>
                <Column>Row2</Column>
              </Row>
            }
          />
        ),
            title = wrapper.find('[data-element="title"]'),
            columns = title.find(Column);

        expect(columns.first().props().children).toEqual('Row1');
        expect(columns.last().props().children).toEqual('Row2');
      });
    });

    describe('when a props title is not passed', () => {
      it('defaults to null', () => {
        const wrapper = shallow(
          <Dialog
            onCancel={ onCancel }
            open
          />
        );
        expect(wrapper.instance().dialogTitle).toBeFalsy();
      });
    });
  });

  describe('render', () => {
    describe('when dialog is open', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Dialog
            onCancel={ onCancel }
            className='foo'
            open
          >

            <Button>Button</Button>
            <Button>Button</Button>
          </Dialog>
        );
      });

      it('renders a parent div with mainClasses attached', () => {
        const dialogNode = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'div')[0];
        expect(dialogNode.className).toEqual('carbon-dialog foo');
      });

      it('renders the dialog', () => {
        expect(instance._dialog).toBeTruthy();
        expect(instance._dialog.classList[0]).toEqual('carbon-dialog__dialog');
      });

      it('closes when the exit icon is click', () => {
        const closeIcon = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-dialog__close');
        TestUtils.Simulate.click(closeIcon);
        expect(onCancel).toHaveBeenCalled();
      });

      it('renders the children passed to it', () => {
        const buttons = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'button');
        expect(buttons.length).toEqual(2);
      });

      describe('when passing a custom size', () => {
        it('adds the size class to the dialog', () => {
          instance = TestUtils.renderIntoDocument(
            <Dialog
              onCancel={ onCancel }
              open
              size='small'
            />
          );

          expect(instance._dialog.classList[1]).toEqual('carbon-dialog__dialog--small');
        });
      });
    });

    describe('when dialog is closed', () => {
      it('renders a parent div with mainClasses attached', () => {
        instance = TestUtils.renderIntoDocument(
          <Dialog open={ false } onCancel={ onCancel } />
        );
        const dialogNode = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'div')[0];
        expect(dialogNode.classList[0]).toEqual('carbon-dialog');
      });
    });
  });

  describe('tags', () => {
    describe('on component', () => {
      const wrapper = shallow(
        <Dialog
          data-element='bar'
          onCancel={ () => {} }
          onConfirm={ () => {} }
          open
          data-role='baz'
        />);

      it('include correct component, element and role data tags', () => {
        rootTagTest(wrapper, 'dialog', 'bar', 'baz');
      });
    });

    describe('on internal elements', () => {
      const wrapper = mount(
        <Dialog
          onCancel={ () => {} }
          onConfirm={ () => {} }
          open
          showCloseIcon
          subtitle='Test'
          title='Test'
        />
      );

      elementsTagTest(wrapper, [
        'close',
        'subtitle',
        'title'
      ]);
    });
  });
});
