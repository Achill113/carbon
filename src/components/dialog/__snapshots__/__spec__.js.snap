// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog dialog headers when a props object title is passed wraps it within a div 1`] = `
<div
  className="carbon-dialog__title"
>
  <Row
    gutter="medium"
  >
    <div
      className="carbon-row carbon-row--gutter-medium carbon-row--columns-2"
    >
      <Column
        key=".0"
      >
        <div
          className="carbon-column"
        >
          Row1
        </div>
      </Column>
      <Column
        key=".1"
      >
        <div
          className="carbon-column"
        >
          Row2
        </div>
      </Column>
    </div>
  </Row>
</div>
`;

exports[`Dialog render when dialog is open has the correct content, tags, elements etc 1`] = `
<Dialog
  ariaRole="dialog"
  autoFocus={true}
  className="foo"
  data-element="bar"
  data-role="baz"
  height="500"
  onCancel={[Function]}
  onConfirm={[Function]}
  open={true}
  showCloseIcon={true}
  size="small"
  subtitle="Test"
  title="Test"
>
  <div
    className="carbon-dialog foo"
    data-component="dialog"
    data-element="bar"
    data-role="baz"
    data-state="open"
  >
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal"
    >
      <TransitionGroup
        childFactory={[Function]}
        component="div"
        transitionAppear={false}
        transitionEnter={true}
        transitionEnterTimeout={500}
        transitionLeave={true}
        transitionLeaveTimeout={500}
        transitionName="modal"
      >
        <div>
          <CSSTransitionGroupChild
            appear={false}
            enter={true}
            enterTimeout={500}
            key=".0"
            leave={true}
            leaveTimeout={500}
            name="modal"
          >
            <div
              aria-describedby="carbon-dialog-subtitle"
              aria-labelledby="carbon-dialog-title"
              className="carbon-dialog__dialog carbon-dialog__dialog--small carbon-dialog__dialog--has-height"
              data-component="dialog"
              data-element="bar"
              data-role="baz"
              key=".0"
              onBlur={[Function]}
              role="dialog"
              style={
                Object {
                  "minHeight": "500px",
                }
              }
              tabIndex={0}
            >
              <div
                className="carbon-dialog__title carbon-dialog__title--has-subheader"
              >
                <Heading
                  divider={true}
                  separator={false}
                  subheader="Test"
                  subtitleId="carbon-dialog-subtitle"
                  title="Test"
                  titleId="carbon-dialog-title"
                >
                  <div
                    className="carbon-heading carbon-heading--has-subheader carbon-heading--has-divider"
                    data-component="heading"
                  >
                    <div
                      className="carbon-heading__header"
                    >
                      <div
                        className="carbon-heading__headers"
                      >
                        <div
                          className="carbon-heading__main-header"
                        >
                          <h1
                            className="carbon-heading__title"
                            data-element="title"
                            id="carbon-dialog-title"
                          >
                            Test
                          </h1>
                        </div>
                        <div
                          className="carbon-heading__subheader"
                          data-element="subtitle"
                          id="carbon-dialog-subtitle"
                        >
                          Test
                        </div>
                      </div>
                    </div>
                  </div>
                </Heading>
              </div>
              <div
                className="carbon-dialog__content"
              >
                <div
                  className="carbon-dialog__inner-content"
                  style={
                    Object {
                      "minHeight": "calc(500px - 40px)",
                    }
                  }
                >
                  <Button
                    as="secondary"
                    disabled={false}
                    size="medium"
                    subtext=""
                    theme="blue"
                  >
                    <button
                      className="carbon-button carbon-button--secondary carbon-button--blue carbon-button--medium"
                      data-component="button"
                      disabled={false}
                    >
                      Button
                    </button>
                  </Button>
                  <Button
                    as="secondary"
                    disabled={false}
                    size="medium"
                    subtext=""
                    theme="blue"
                  >
                    <button
                      className="carbon-button carbon-button--secondary carbon-button--blue carbon-button--medium"
                      data-component="button"
                      disabled={false}
                    >
                      Button
                    </button>
                  </Button>
                </div>
              </div>
              <Icon
                bgSize="small"
                className="carbon-dialog__close"
                data-element="close"
                onBlur={[Function]}
                onClick={[Function]}
                tabIndex="0"
                type="close"
              >
                <span
                  className="carbon-icon carbon-dialog__close icon-close"
                  data-component="icon"
                  data-element="close"
                  onBlur={[Function]}
                  onClick={[Function]}
                  tabIndex="0"
                  type="close"
                />
              </Icon>
            </div>
          </CSSTransitionGroupChild>
        </div>
      </TransitionGroup>
    </CSSTransitionGroup>
    <CSSTransitionGroup
      component="div"
      transitionAppear={false}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal-background"
    >
      <TransitionGroup
        childFactory={[Function]}
        component="div"
        transitionAppear={false}
        transitionEnter={true}
        transitionEnterTimeout={500}
        transitionLeave={true}
        transitionLeaveTimeout={500}
        transitionName="modal-background"
      >
        <div>
          <CSSTransitionGroupChild
            appear={false}
            enter={true}
            enterTimeout={500}
            key=".0"
            leave={true}
            leaveTimeout={500}
            name="modal-background"
          >
            <div
              className="carbon-modal__background"
              key=".0"
            />
          </CSSTransitionGroupChild>
        </div>
      </TransitionGroup>
    </CSSTransitionGroup>
  </div>
</Dialog>
`;
